### PR TITLE
[MRG+1] DOC: switch pylab example `mri_with_eeg.py` to OO interface + cosmetick fixes

### PR DESCRIPTION
**Pull request by afvincent**
The first commit switch the script to pyplot OO interface. The next ones perform some minor cosmetick adjustements, either in the plot layout or the doctsrings and the comments.

If the added MultipleLocator is too much, then I think that at least one should not force negative xticks as former l. 32 did: they are no negative values in the MRI data.

Additional remark: I am not a big fan of the 
```python
if 1:  # <Description>
```
blocks. I wonder if simple
```python
"""
<Description>
"""
```
would not be as well OK visually and more semantically correct.

### CODE REVIEW

File: examples/pylab_examples/mri_with_eeg.py [hunk 1]
```diff
@@ -1,77 +1,79 @@
...
-import matplotlib.cbook as cbook
-# I use if 1 to break up the different regions of code visually
+from matplotlib.ticker import MultipleLocator
 
-if 1:   # load the data
-    # data are 256x256 16 bit integers
+# NB: one uses "if 1:" to break up the different regions of code visually
+fig = plt.figure("MRI_with_EEG")
+
+if 1:   # Load the data
```
Comment (NelleV) [inline examples/pylab_examples/mri_with_eeg.py]: I really dislike this syntay. Can we get rid of this? It is using python code for documentation purposes.
Having comments should be enough to break the different regions of the code visually.

Comment (afvincent) [inline examples/pylab_examples/mri_with_eeg.py]: I pushed a new commit that get rid of this idiom.

File: examples/pylab_examples/mri_with_eeg.py [hunk 2]
```diff
@@ -14,66 +14,75 @@
...
-
-    lines = LineCollection(segs, offsets=offsets, transOffset=None)
-    ax2.add_collection(lines)
-
-    # Set the yticks to use axes coords on the y axis
-    ax2.set_yticks(ticklocs)
-    ax2.set_yticklabels(['PG3', 'PG5', 'PG7', 'PG9'])
-
-    ax2.set_xlabel('Time (s)')
+"""
```
Comment (NelleV) [inline examples/pylab_examples/mri_with_eeg.py]: I am sorry, but I still don't like this. This is also python code,  meant for a specific purpose. Can you switch to comments?

**Comment by Kojoley** [comment]
@afvincent @NelleV maybe it is better to convert these branches to functions like `def plot_xxx(fig):`?

**Comment by NelleV** [comment]
@Kojoley that would make the example slightly harder to follow. I am pretty sure it would be quite easy to make the blocks stand out just with comments and approrpiate blank lines. We have many examples that are more complex than this one in the documentation, and are yet very readable.

**Comment by afvincent** [comment]
@Kojoley In this particular case, I am not sure that converting everything into functions is worth the extra overhead, as there is nothing really repeating itself among the different parts of the script.

@NelleV I switched to simple comment lines: I hope it is closer to what you had in mind.

By the way, here is what the picture looks like with this PR:
mri_with_eeg
See here for the current picture: [http://matplotlib.org/devdocs/examples/pylab_examples/mri_with_eeg.html](http://matplotlib.org/devdocs/examples/pylab_examples/mri_with_eeg.html). The differences are about the upper right panel (it had a useless negative x-axis part and the x-label was hidden by the lower panel).

**Comment by NelleV** [comment]
That looks really cool :)

File: examples/pylab_examples/mri_with_eeg.py [hunk 3]
```diff
@@ -1,77 +1,79 @@
...
+# Plot the MRI image
+ax0 = fig.add_subplot(2, 2, 1)
+ax0.imshow(im, cmap=cm.gray)
+ax0.axis('off')
+
+# Plot the histogram of MRI intensity
+ax1 = fig.add_subplot(2, 2, 2)
+im = np.ravel(im)
+im = im[np.nonzero(im)]  # Ignore the background
+im = im / (2**15)  # Normalize
```
Comment (QuLogic) [inline examples/pylab_examples/mri_with_eeg.py]: So this is how it was originally, but I'm not sure if it's right. The intensities go over 1 and a `uint16` should have a maximum of 2**16, but maybe this just how MRIs are defined? I don't know, but it seems odd.

File: examples/pylab_examples/mri_with_eeg.py [hunk 4]
```diff
@@ -1,77 +1,79 @@
...
+ax1.xaxis.set_major_locator(MultipleLocator(0.5))
+ax1.set_yticks([])
+ax1.set_xlabel('Intensity')
+ax1.set_ylabel('MRI density')
+
+# Load the EEG data
+numSamples, numRows = 800, 4
+eegfile = cbook.get_sample_data('eeg.dat', asfileobj=False)
+print('Loading EEG %s' % eegfile)
+data = np.fromstring(open(eegfile, 'rb').read(), float)
```
Comment (QuLogic) [inline examples/pylab_examples/mri_with_eeg.py]: Leaky `open`; can use `np.fromfile(eegfile, float)` instead

File: examples/pylab_examples/mri_with_eeg.py [hunk 5]
```diff
@@ -1,77 +1,79 @@
...
-
-    xlabel('time (s)')
-
-show()
+from matplotlib.ticker import MultipleLocator
+
+fig = plt.figure("MRI_with_EEG")
+
+# Load the MRI data (256x256 16 bit integers)
+dfile = cbook.get_sample_data('s1045.ima.gz')
```
Comment (QuLogic) [inline examples/pylab_examples/mri_with_eeg.py]: This file should be closed.

File: examples/pylab_examples/mri_with_eeg.py (hunk 3 above)
Comment (NelleV) [inline examples/pylab_examples/mri_with_eeg.py]: ping @matthew-brett our resident expert on MRI :)

File: examples/pylab_examples/mri_with_eeg.py (hunk 5 above)
Comment (NelleV) [inline examples/pylab_examples/mri_with_eeg.py]: `get_sample_data` should probably be deprecated and replaced with something with a sane interface.
In any case, python will close that file automatically at the end of the script without problems considering it is a read-only file.

File: examples/pylab_examples/mri_with_eeg.py (hunk 3 above)
Comment (afvincent) [inline examples/pylab_examples/mri_with_eeg.py]: I agree that this normalization seemed weird to me too but I did not feel comfortable changing it to (2**16 - 1) as I do not know anything about MRI. PS: anyway, currently there are no units for the MRI...

**Comment by afvincent** [comment]
While waiting for more info on what to do about the normalization, the commit c8388f9 takes care of the leaky `open` that were present.

**Comment by afvincent** [comment]
@NelleV Looking at the files in the `pylab_examples`, there is another (simpler) example `mri_demo.py` with issues similar to `mri_with_eeg.py` (incomplete OO interface, not close file, room for docstring improvement, etc.). As I would not like to mess with the "[MRG + 1]" and GH "approved review" process,should I open a new dedicated PR or is it possible to add a "corrected" version of `mri_demo.py` to the current PR?

**Comment by NelleV** [comment]
Either is fine to me. I can also review it and readapt my +1 whether the patch seems satisfying or not.

I have started on the integration of sphinx-gallery yesterday, so I may have new specific requests on how to document examples (It is basically a docstring with a title, and a description - nothing fancy).

File: examples/pylab_examples/mri_with_eeg.py [hunk 6]
```diff
@@ -1,77 +1,80 @@
-"""
-This now uses the imshow command instead of pcolor which *is much
-faster*
+"""Displays a set of subplots with an MRI image, its intensity histogram and
+some EEG traces.
+
+NB: ones uses the imshow command instead of pcolor which *is much faster*.
```
Comment (afvincent) [inline examples/pylab_examples/mri_with_eeg.py]: You were right (see 6721163). I have removed it.

**Comment by afvincent** [comment]
Finally I have normalized the MRI data in `mri_with_eeg.py` with the np.uint16 max value (dfc9cca) and adjusted the x-ticks and x-label accordingly. If somebody more in this field says it was a mistake, I will reverse it.

@NelleV  The next two commits (db3d206 and 96ff817) fix the similar example (`mri_demo.py`) in a similar fashion to the MRI subplot in `mri_with_eeg.py`. If the docstrings of these two examples have to be enhanced, just let me know.

**Comment by NelleV** [comment]
@afvincent I am not happy with what sphinx-gallery requires (there is a small incoherence IMO between the way you write a title) so I am opening a bug on the sphinx-gallery issue tracker.
Basically, it requires a title and a description *unless* the title is written in rst format, ie:

```
=====
title
=====
```

If the title is written in rst format, then sphinx-gallery also displays it as the figure name in the gallery (instead of the file name as it is right now in our gallery).
If you are up to write a docstring like the following, that'd be awesome but it requires some work of understanding why the plot has been added to the gallery, etc
https://github.com/sphinx-gallery/sphinx-gallery/blob/master/examples/just_code.py 

Last element, we'll have to rename examples that plot a figure into `plot_something.py` for sphinx-gallery to display the plot in the gallery. So you can either start doing it in this PR, or I'll do a batch rename later on.

**Comment by NelleV** [comment]
Never mind… It does seem to require a header and a text at all times. I don't know how I manage to compile the documentation before…

**Comment by efiring** [comment]
It's a little late, but I will raise the question anyway: is there any point in keeping mri_demo, given that its contents are repeated in mri_with_eeg?

**Comment by NelleV** [comment]
@efiring That's an excellent questions. I've going through the examples lately, and there are *a lot* of repeats with slight variations. Writing up a bit of documentation on these examples will probably help identifying duplicates and reduce our examples size.

**Comment by tacaswell** [comment]
I really wish that the old comments got folded again :disappointed:

**Comment by tacaswell** [comment]
Backported to v2.x as 17fab80

**Pull request merged**